### ISSUE TITLE
Oj.load memory leak for big files

### ISSUE THREAD

**Issue by michaelpulis** [ohler55/oj#679]
```Oj.load``` is leaking memory for big json files (32 mb in this case). Consider the following code, where ```outRam``` is a function that outputs the RAM being used by the current process, and ```readFile(...)``` is a function that reads the string from the file. This reading function has been tested independently and it does not leak any memory.
```ruby
puts "Starting"
outRAM

file = readFile('big_file.json')
test_repo = Oj.load(file)

file = nil
test_repo = nil

GC.start
outRAM
```

This is the output of the command:
```
Starting
RAM: [total 54564K]
RAM: [total 303740K]
```

The RAM used by the ``` Oj.load ``` method is never freed until the process is killed.

**Comment by ohler55** [comment]
If you can provide the readFile and outRAM functions I can test myself. One thing to remember though. Some memory on the heap is returned differently than Ruby objects. The best way to make sure there is no leak is to run in a loop and see if the size remains the same or grows. A single call is not an accurate way to determine if there is a leak as apposed to reserved space that will be reused.

**Comment by michaelpulis** [comment]
The ```readFile``` method is as follows:
```ruby
def readFile
   File.read("/path/to/file")
end
```

The ```outRAM``` function is as follows
```ruby
def outRAM
   puts `pmap #{Process.pid} | grep total`
end
```

Running this in a loop leads to the RAM increasing for the first 1-2 iterations, and then stabilising. How would I clear the memory that is on the heap, if this is the issue? ```GC.start``` does not clear this memory, thus I assumed that it was a memory leak.

**Comment by ohler55** [comment]
Oj as well as some other gems make use of memory that is not used for Ruby objects. A GC does not clear that memory. That memory is kept in the heap which may be freed at some point. There are also cases where it will not be. For example if there is a cache of hash keys that is not freed since it may be used again.

By running multiple iterations the gem is "warmed up" and should not grow unless something new comes along like a larger file or maybe some new keys if those are being cached.

So basically, from your tests it sounds like there is not a memory leak.

**Comment by michaelpulis** [comment]
Can this heap memory be freed?

**Comment by mblumtritt** [comment]
If you try to use `GC.start` keep in mind that all variables used _in current context_ are normally not freed. It may help - but not in every case! - to allocate the memory in a function call or something:

```ruby

def test_oj
   file = readFile('big_file.json')
   test_repo = Oj.load(file)
   nil
end

puts "Starting"
outRAM

test_oj

GC.start
outRAM
```

**Comment by michaelpulis** [comment]
@mblumtritt This does not clear the heap. In the case it is being used, the big JSON being loaded is being changed constantly, thus all the heap space does add up. Can this be cleared or avoided @ohler55 ?

**Comment by mblumtritt** [comment]
What about using `Oj.load_file`? I don't know the implementation details, but this may not read all content at once into memory. What do you say @ohler55?

**Comment by ohler55** [comment]
By loading the file you have effectively allocated space for the contents of the file. That content takes of heap space. Depending on the OS it may or not be recovered but as you noticed the space is reused on subsequent parser calls.

Using the Oj.load_file will not change that. There is a new parser in the works that does parse incrementally and that will keep the memory use down depending on how it is used. If many Ruby objects are created they still make use of ram. The new parser is also faster than the current Oj parser with some other interesting features. I expect a release near the end of the month.

**Comment by michaelpulis** [comment]
The file being loaded is around 30mb, but the permanent increase in memory usage is of around 300mb, as I showed in the output block of my first post in this thread. Can this amount be reduced ?

**Comment by mblumtritt** [comment]
@michaelpulis 
You have here a general Ruby dependent scenario but not an Oj specific. I used in a similar situations `GC.compact` and ensured that all related memory was not longer part of the current context…

**Comment by ohler55** [comment]
If you want to find the base memory try running your scenario with a trivial file that's only a few bytes. That should load in all the dependencies. Then run with the larger file for several iterations and see what the change is over time.
